src/browser/file-recovery-service.js

```diff
@@ -0,0 +1,65 @@
+'use babel'
+
+import {ipcMain} from 'electron'
+import crypto from 'crypto'
+import Path from 'path'
+import fs from 'fs-plus'
+
+export default class FileRecoveryService {
+  constructor (recoveryDirectory) {
+    this.recoveryDirectory = recoveryDirectory
+    this.recoveryPathsByWindowAndFilePath = new WeakMap()
+    this.crashListeners = new WeakSet()
+  }
+
+  start () {
+    ipcMain.on('will-save-path', this.willSavePath.bind(this))
+    ipcMain.on('did-save-path', this.didSavePath.bind(this))
+  }
+
+  willSavePath (event, path) {
+    if (!fs.existsSync(path)) {
+      // Unexisting files won't be truncated/overwritten, and so there's no data to be lost.
+      return
+    }
+
+    const window = event.sender
+    const recoveryFileName = crypto.randomBytes(5).toString('hex')
+    const recoveryPath = Path.join(this.recoveryDirectory, recoveryFileName)
+    fs.writeFileSync(recoveryPath, fs.readFileSync(path))
```

Comment:
I am not sure about flushing, as I don't think it's enough to ensure the file was persisted. From `man 2 fsync`:

```
     Specifically, if the drive loses power or the OS crashes, the application may find that only some or none of their data was writ-
     ten.  The disk drive may also re-order the data so that later writes may be present, while earlier writes are not.

     This is not a theoretical edge case.  This scenario is easily reproduced with real world workloads and drive power failures.

     For applications that require tighter guarantees about the integrity of their data, Mac OS X provides the F_FULLFSYNC fcntl.  The
     F_FULLFSYNC fcntl asks the drive to flush all buffered data to permanent storage.  Applications, such as databases, that require a
     strict ordering of writes should use F_FULLFSYNC to ensure that their data is written in the order they expect.  Please see
     fcntl(2) for more detail.
```

So we might use `fcntl` but that wouldn't be cross-platform and I am not sure that investing time in flushing data to disk is actually worth it. What I like about this solution is that, worst-case scenario, users can find the contents of their file under `~/.atom/recovery` after rebooting their machine when e.g. the system crashes.